src/lib/clk.c

```diff
@@ -92,48 +82,31 @@ void clock_set_freq(int clock, uint32_t hz)
 	notifier_event(&notify_data);
 
 	if (!clk_info->set_freq ||
-	    clk_info->set_freq(clock, idx) == 0) {
+	    clk_info->set_freq(clock, idx) == 0)
 		/* update clock frequency */
-		clk_pdata->clk[clock].freq = clk_info->freqs[idx].freq;
-		clk_pdata->clk[clock].ticks_per_msec =
-			clk_info->freqs[idx].ticks_per_msec;
-	}
+		clk_info->current_freq_idx = idx;
 
 	/* tell anyone interested we have now changed freq */
 	notify_data.message = CLOCK_NOTIFY_POST;
 	notifier_event(&notify_data);
 
-	spin_unlock_irq(clk_pdata->clk[clock].lock, flags);
+	spin_unlock_irq(clk_info->lock, flags);
 }
 
 uint64_t clock_ms_to_ticks(int clock, uint64_t ms)
 {
-	return clk_pdata->clk[clock].ticks_per_msec * ms;
+	struct clock_info *clk_info = &clocks[clock];
+
+	return clk_info->freqs[clk_info->current_freq_idx].ticks_per_msec * ms;
 }
 
 void platform_timer_set_delta(struct timer *timer, uint64_t ns)
 {
+	struct clock_info *clk_info = &clocks[PLATFORM_DEFAULT_CLOCK];
+	uint32_t ticks_per_msec =
+		clk_info->freqs[clk_info->current_freq_idx].ticks_per_msec;
 	uint64_t ticks;
 
-	ticks = clk_pdata->clk[PLATFORM_DEFAULT_CLOCK].ticks_per_msec /
-		1000 * ns / 1000;
+	ticks = ticks_per_msec / 1000 * ns / 1000;
```

Comment:
This isn't something that this patch adds or modifies, but (taking out my K&R copy), I think what is happening above is, that first `ticks_per_msec / 1000` is calculated, where both are 32-bit values, so, it's 32-bit arithmetic. I'm not sure what typical `ticks_per_msec` values are, but I guess some clocks can be in low MHz, in which case that would be just a few thousands, so, that division would introduce some not insignificant rounding down. Then `result32 * ns` will be calculated, where `result32` is 32-bit, but `ns` is 64-bit. So, before the multiplication `result32` will be converted to 64 bits too. Then `result64 / 1000` is calculated with 1000 promoted to 1000LL. So, we're doing 2 64-bit and one 32-bit operation and losing some precision. We could instead just do `ticks_per_msec * ns / 1000000` which would be 2 64-bit operations and both save time and improve precision...

Reply:
I guess @tlauda just wanted to keep the code logically the same, but you are right that what you proposed is optimization here